Cody: Fix hallucination detection

deepak2431 opened the pull request:
This PR attempted to fix #52023, and #52909.

Changes done:
1. Added validation for API pattern
2. Added validation for the git refs pattern
3. Fix bug where the status of file ending with "/" was returning undefined.

## Test plan

Added tests for the changes made.

deepak2431 commented:
@philipp-spiess Can I have a review, please?

philipp-spiess commented on client/cody-shared/src/hallucinations-detector/index.ts:
```diff
@@ -163,6 +163,18 @@ function isFilePathLike(fullMatch: string, pathMatch: string): boolean {
-    // TODO: we can do further validation here.
+
+    // check for API endpoints
+    const apiRegex = new RegExp('^/[^\\s]*[./:][^\\s]*$')
```
I don't really understand what this regex is trying to construct. Do you mind explaining the intention here?

philipp-spiess replied:
I think that this might be a bit too unrestrictive for this kind of check. e.g. `/some/path` is matched by it but it might also be a valid directory?

deepak2431 replied:
This regex is trying to check basically for the strings starting with `/` and have `/:` and similar patterns.  So, API endpoints like `/api/product/:id` can be matched.

To your next point, I agree that `/customer/products` can also be a directory but this can also be an endpoint route if it doesn't have the `/api` keyword.

Also, I would like to mention that most directory strings in the token are of that format  `/some/path/`, so I think that won't be an issue with this test.

LMK what do you think?

philipp-spiess replied:
I see but I don't think the current regex is good enough at detecting it and it has a lot of false positives, e.g. `/package.json` is matched.

Should we update the logic so that we require it to have at least one _named param_ for this rule? so we match the `/product/:id` use case but not `/product/123`? 

We can do that by updating the regex to something like this:

```
/^\/[^\s]*\/:[^\s]*$/
```

deepak2431 replied:
Yeah, agreed with you here. Let me update the regex. Thanks for the suggestion!

deepak2431 commented:
@philipp-spiess I have updated the regex to a new one, as the one you suggested was getting updated due to ESLint. Added some tests to make sure it's good.

philipp-spiess commented:
@deepak2431 Yeah I think this should be good too 😎 Let's run the test on CI
